### ISSUE TITLE
Functional API suggestion

### ISSUE THREAD

**Issue by jurra** [szkuds/TUD_LBM#41]
@szkuds and @khoefnagel, I have been working lately on this, please have a look before our meeting to see if it makes sense. 

Before implementing this is important to have tests in place, and make sure that it runs before we refactor. I would also like to discuss the test I have been working on that is an integration test. This would be an anchor point where we know everything is working and tested before we move on to refactor. 

# Why the Functional API Matters: 

## The Problem: Old Class-Based Approach

**Before (what we're replacing):**
```python
# Mutable state hidden in object
class SinglePhaseSimulation:
    def __init__(self, config):
        self.lattice = create_lattice(config)
        self.f = None
        self.rho = None
        self.u = None
        self.t = 0
        
    def initialise(self):
        self.f = initialise_standard(...)  # <-- MUTATION
        self.rho = jnp.ones((nx, ny, 1, 1))  # <-- MUTATION
        self.t = 0
        
    def run_step(self):
        self.f = collision(self.f, self.feq, tau)  # <-- MUTATION
        self.f = streaming(self.f)  # <-- MUTATION
        self.rho, self.u = macroscopic(self.f)  # <-- MUTATION
        self.t += 1  # <-- MUTATION
        return StepResult(f=self.f, rho=self.rho, u=self.u, t=self.t)

# Usage (you never know what's in self anymore)
sim = SinglePhaseSimulation(config)
sim.initialise()
for step in range(nt):
    result = sim.run_step()  # What happens to sim state?
```

**Problems with this approach:**

1. **Hidden State (No Transparency)**
   - You create `sim` and call methods
   - You don't know what state changed until you look inside
   - After `run_step()`, which attributes were modified?
   - Hard to debug: "where did this bug come from?"

2. **Non-deterministic Behavior**
   - Same input might produce different output if `sim` state is unexpected
   - Testing requires careful setup/teardown of object state
   - Parallel execution is dangerous (multiple threads accessing `sim.f`)

3. **JAX Cannot JIT Compile Efficiently**
   - JAX's `jax.lax.scan` requires pure functions
   - Cannot JIT a method like `self.run_step()` because JAX doesn't know what `self` contains
   - Results in repeated JIT compilations, cache bloat, performance loss

4. **Difficult to Test**
   - Must instantiate entire `sim` object for each test
   - Cannot easily mock or stub internal state
   - Integration tests become entangled with implementation details

5. **Cannot Compose or Parallelize**
   - Cannot run two simulations without creating two objects
   - Cannot run one simulation in multiple processes (state is not serializable)
   - Cannot use functional composition patterns

---

## The Solution: Functional API with Pure Functions & Immutable Data

**After (what we're implementing):**
```python
# Step 1: Create immutable config
config = SimulationConfig(
    sim_type="single_phase",
    nx=5, ny=32,
    lattice="d2q9",
    collision_scheme="bgk",
    tau=0.8,
    force_config={"type": "gravity_multiphase", "force_g": 1e-6, "inclination_angle_deg": 90.0}
)

# Step 2: Build setup (one-time, immutable computation)
setup = build_setup(config)
# Returns: SimulationSetup(lattice, masks, templates, diff_ops)
# All created once, never changes

# Step 3: Initialize state (pure function)
state = init_state(setup)
# Returns: State(f, rho, u, t, force, force_ext, h, wetting)
# State is a NamedTuple — immutable

# Step 4: Run simulation (pure function over time)
final_state, history = run(setup, state, nt=5000)
# Input: (setup, state, nt)
# Output: (final_state, history)
# NO SIDE EFFECTS

# Usage (crystal clear what happens)
for step_idx in range(nt):
    state = step_single_phase(setup, state)
    # Returns NEW state, doesn't modify old one
```

#### 1. **Pure Functions = Deterministic + Transparent**

```python
# Pure function: same input ALWAYS produces same output
def collision_step(f: Array, feq: Array, tau: float) -> Array:
    """Transform populations toward equilibrium.
    
    INPUT: f (nonequilibrium), feq (equilibrium), tau (relaxation time)
    OUTPUT: f_col (post-collision populations)
    SIDE EFFECTS: None
    """
    return f - (f - feq) / tau

# You know exactly what it does:
# - Takes three inputs
# - Returns one output
# - Changes nothing else in the world
# - Same input → same output (every single time)

# vs. before:
class CollisionOperator:
    def __init__(self, tau):
        self.tau = tau
        self.cache = {}  # Hidden state!
        
    def apply(self, f, feq):
        # Does it use self.cache? Modify self? Who knows?
        return f - (f - feq) / self.tau
```

**Research benefit:** Reproducibility. Pure functions are inherently reproducible. No hidden state, no surprises.

#### 2. **Immutable Data = Safe Composition + History Tracking**

```python
# Immutable state — a NamedTuple
state_0 = State(f=f_init, rho=rho_init, u=u_init, t=0, ...)

# Take a step → get NEW state, old one unchanged
state_1 = step_single_phase(setup, state_0)

# You can keep the history!
history = [state_0, state_1, state_2, ...]

# Compare any two timesteps without rerunning simulation
u_difference = state_10.u - state_5.u

# vs. before:
sim.run_step()  # Overwrites sim.u
sim.run_step()  # Overwrites sim.u again
# u from step 5 is lost forever
```

**Research benefit:** You can analyze flow evolution, compute statistics over time, visualize trajectories, all without storing intermediate states separately.

#### 3. **JAX JIT Compilation: 10-100x Speedup**

```python
# JAX can JIT-compile pure functions
@jax.jit
def step_jit(setup, state):
    """Pure function — JAX loves this!"""
    rho, u = compute_macroscopic(state.f, setup.lattice)
    feq = compute_equilibrium(rho, u, setup.lattice)
    f_col = collision_op(state.f, feq, tau)
    f_stream = streaming_op(f_col, setup.lattice)
    f_bc = boundary_op(f_stream, f_col, setup.bc_masks)
    return State(f=f_bc, rho=rho, u=u, ...)

# Run 5000 steps with ONE JIT compilation
for _ in range(5000):
    state = step_jit(setup, state)  # <-- JIT compiled once, runs fast

# Can even use jax.lax.scan for auto-vectorization:
final_state, history = jax.lax.scan(
    lambda state, _: (step_jit(setup, state), state),
    state_0,
    jnp.arange(5000)
)
# All 5000 steps compiled as a single unit!
```

**Research benefit:** JAX can compile to GPU/TPU. Your simulation runs 10-100x faster without changing physics code.

#### 4. **Testing: Pure Functions Are Easy to Test**

```python
# Test collision operator in isolation
def test_collision_conserves_mass():
    f = jnp.ones((5, 32, 1, 9)) / 9  # Rest equilibrium
    feq = equilibrium(rho=1.0, u=(0, 0), lattice=d2q9)
    
    f_col = collision_op(f, feq, tau=0.8)
    
    # Pure function: no setup needed, no teardown
    assert jnp.allclose(jnp.sum(f), jnp.sum(f_col))
    # Just assert the physics guarantee

# vs. before:
def test_collision():
    sim = SinglePhaseSimulation(config)  # Create whole object
    sim.initialise()  # Set up state
    old_mass = jnp.sum(sim.f)
    sim.run_step()  # Hope it only changed f?
    new_mass = jnp.sum(sim.f)
    assert jnp.allclose(old_mass, new_mass)
    # What if run_step also modified rho? Or t? Or something else?
```

**Research benefit:** Unit tests run fast, are easy to write, and don't accidentally test implementation details.

#### 5. **Swappable Components = Easy Experimentation**

```python
# BEFORE: To try different collision model, rewrite everywhere
class BGK_Simulation:
    def run_step(self):
        self.f = collide_bgk(self.f, self.feq, self.tau)
        
class MRT_Simulation:
    def run_step(self):
        self.f = collide_mrt(self.f, self.feq, self.k_diag)
        
# Now you have TWO classes to maintain!

# AFTER: Pure function approach
def run(setup, state, collision_op, nt):
    """Generic run — doesn't care which collision."""
    for _ in range(nt):
        rho, u = compute_macroscopic(state.f, setup.lattice)
        feq = compute_equilibrium(rho, u, setup.lattice)
        f_col = collision_op(state.f, feq, tau)  # <-- Operator passed in
        f_stream = streaming_op(f_col, setup.lattice)
        state = state.replace(f=f_stream, rho=rho, u=u)
    return state

# Swap operators without changing run():
state_bgk = run(setup, state_0, collision_op=collide_bgk, nt=5000)
state_mrt = run(setup, state_0, collision_op=collide_mrt, nt=5000)
state_custom = run(setup, state_0, collision_op=collide_my_scheme, nt=5000)

# Compare results: Does MRT give different stability? Different accuracy?
```

**Research benefit:** Experiment with different numerical schemes without rewriting entire codebase.

---

## From Functional API to Responsibility Driven Design

The functional API enables **Responsibility Driven Design**:

```
OLD (class-based):
┌────────────────────────────┐
│ SinglePhaseSimulation      │
│  - initialise()            │  Everything tangled together
│  - run_step()              │  Hard to understand responsibility
│  - .f, .rho, .u, .t (mut.) │  Hidden state, hidden mutations
└────────────────────────────┘

NEW (functional):
┌──────────────────────────────────────────────────┐
│ Pure Function Composition                        │
│                                                  │
│ config → build_setup(config) → setup             │
│            (one-time setup)                      │
│                                                  │
│ setup, state → step(setup, state) → new_state    │
│   - MacroscopicOperator: extract moments         │
│   - EquilibriumOperator: compute f_eq            │
│   - CollisionOperator: apply collision           │
│   - StreamingOperator: propagate                 │
│   - BoundaryOperator: enforce BC                 │
│                                                  │
│ Each operator has ONE responsibility             │
│ Each step returns NEW state (doesn't mutate)     │
│ Easy to test, compose, parallelize, optimize     │
└──────────────────────────────────────────────────┘
```

---

## Connection to Physical Understanding

**As a researcher, why should you care?**

Your simulation is a **sequence of well-defined transformations**:

1. **State** = {populations f, density ρ, velocity u, time t}
2. **Each step** = {State → State} (a pure function)
3. **Simulation** = Composing steps: State₀ → State₁ → State₂ → ... → StateN

When you write it this way:
- You can reason about what each step does
- You can prove properties hold (e.g., "collision conserves mass")
- You can test each step independently
- You can swap steps and compare results
- You can JIT-compile for speed
- You can parallelize across multiple initial conditions
- **You can understand your own code 6 months later**

---

## The Poiseuille Test as Validation

The Poiseuille integration test validates this entire pipeline:

```python
# 1. Create config (immutable specification)
config = SimulationConfig(...)

# 2. Build setup (immutable compiled pipeline)
setup = build_setup(config)

# 3. Initialize state (immutable starting point)
state = init_state(setup)

# 4. Run simulation (5000 pure function applications)
final_state, _ = run(setup, state, nt=5000)

# 5. Verify physics (u matches analytical parabola)
u_analytical = ...
u_computed = final_state.u
assert L2_error(u_computed, u_analytical) < 0.02
```

**Why this matters:** If the Poiseuille test passes, we know:
- Pure functions compose correctly
- Immutable data flows through pipeline correctly
- Numerical scheme is stable and accurate
- Physics is preserved
- Code is ready for refactoring (test won't break)

---

## Summary: Why Functional API Wins

| Aspect | Class-Based (Old) | Functional (New) |
|--------|-------------------|-----------------|
| **Predictability** | Hidden mutations | Pure functions — deterministic |
| **Testing** | Complex setup/teardown | Unit test any function |
| **JAX JIT** | Cannot JIT efficiently | Compiles to GPU/TPU |
| **Performance** | ⚠️ Slow (Python loops) | ✅ 10-100x faster (JIT) |
| **Experimentation** | Rewrite for each variant | Swap operators at runtime |
| **Parallelization** | State unsafe to share | Immutable data parallelize easily |
| **Understanding** | Requires tracing code | Logic visible in data flow |
| **Reproducibility** | Hidden state breaks reproducibility | Pure functions guarantee reproducibility |
| **Composability** | Cannot easily compose steps | Trivial to compose, reuse, combine |

**Bottom line:** Functional API + immutable data = clearer physics, faster code, easier research.

**Comment by jurra** [comment]
I realize now that this has been implemented at least partially. Nevertheless is good to raise it.
